File: src/gateway/internal/pipeline/cycle.go
package pipeline

import (
	"time"

	"github.com/flamegreat/etradie/src/gateway/internal/constants"
	"github.com/flamegreat/etradie/src/gateway/internal/models"

	"crypto/rand"
	"encoding/hex"
)

// CycleTracker tracks the lifecycle of a single analysis cycle.
type CycleTracker struct {
	cycleID        string
	traceID        string
	status         constants.CycleStatus
	phase          constants.CyclePhase
	outcome        constants.CycleOutcome
	startedAt      time.Time
	startMono      time.Time
	phaseStartMono time.Time
	completedAt    *time.Time
	err            string
	errStage       string
	phaseDurations map[string]float64
}

// NewCycleTracker creates a tracker for a new analysis cycle.
func NewCycleTracker(traceID string) *CycleTracker {
	if traceID == "" {
		traceID = generateID()
	}
	now := time.Now()
	return &CycleTracker{
		cycleID:        generateID(),
		traceID:        traceID,
		status:         constants.StatusRunning,
		phase:          constants.PhaseInitializing,
		startedAt:      now.UTC(),
		startMono:      now,
		phaseStartMono: now,
		phaseDurations: make(map[string]float64),
	}
}

// CycleID returns the unique cycle identifier.
func (t *CycleTracker) CycleID() string { return t.cycleID }

// TraceID returns the distributed trace identifier.
func (t *CycleTracker) TraceID() string { return t.traceID }

// Phase returns the current pipeline phase.
func (t *CycleTracker) Phase() constants.CyclePhase { return t.phase }

// Status returns the current cycle status.
func (t *CycleTracker) Status() constants.CycleStatus { return t.status }

// ElapsedMs returns milliseconds since cycle start.
func (t *CycleTracker) ElapsedMs() float64 {
	return float64(time.Since(t.startMono).Milliseconds())
}

// TransitionTo records a phase transition and the duration of the previous phase.
func (t *CycleTracker) TransitionTo(phase constants.CyclePhase) {
	now := time.Now()
	prevDurationMs := float64(now.Sub(t.phaseStartMono).Milliseconds())
	t.phaseDurations[t.phase.String()] = prevDurationMs
	t.phase = phase
	t.phaseStartMono = now
}

// Complete marks the cycle as completed with the given outcome.
func (t *CycleTracker) Complete(outcome constants.CycleOutcome) {
	now := time.Now()
	prevDurationMs := float64(now.Sub(t.phaseStartMono).Milliseconds())
	t.phaseDurations[t.phase.String()] = prevDurationMs

	t.phase = constants.PhaseCompleted
	t.status = constants.StatusCompleted
	t.outcome = outcome
	nowUTC := now.UTC()
	t.completedAt = &nowUTC
}

// Fail marks the cycle as failed.
func (t *CycleTracker) Fail(errMsg string, stage string, timedOut bool) {
	now := time.Now()
	prevDurationMs := float64(now.Sub(t.phaseStartMono).Milliseconds())
	t.phaseDurations[t.phase.String()] = prevDurationMs

	t.phase = constants.PhaseFailed
	if timedOut {
		t.status = constants.StatusTimedOut
	} else {
		t.status = constants.StatusFailed
	}
	t.outcome = constants.OutcomePipelineError
	t.err = errMsg
	t.errStage = stage
	nowUTC := now.UTC()
	t.completedAt = &nowUTC
}

// ToState returns a snapshot of the current cycle state.
func (t *CycleTracker) ToState() *models.CycleState {
	nowUTC := time.Now().UTC()
	durations := make(map[string]float64, len(t.phaseDurations))
	for k, v := range t.phaseDurations {
		durations[k] = v
	}
	return &models.CycleState{
		CycleID:          t.cycleID,
		TraceID:          t.traceID,
		Status:           t.status,
		Phase:            t.phase,
		Outcome:          t.outcome,
		StartedAt:        t.startedAt,
		PhaseStartedAt:   &nowUTC,
		CompletedAt:      t.completedAt,
		Error:            t.err,
		ErrorStage:       t.errStage,
		PhaseDurationsMs: durations,
	}
}

func generateID() string {
	b := make([]byte, 16)
	_, _ = rand.Read(b)
	return hex.EncodeToString(b)
}

File: src/backup/gateway/pipeline/cycle.py
"""Analysis cycle state tracker.

Pure state machine that tracks phase transitions, timing, and outcomes.
No I/O - only state management and duration recording.
"""

from __future__ import annotations

import time
from datetime import UTC, datetime
from typing import Optional
from uuid import uuid4

from gateway.constants import CycleOutcome, CyclePhase, CycleStatus
from gateway.context.models import CycleState


class CycleTracker:
    """Tracks the lifecycle of a single analysis cycle."""

    def __init__(self, trace_id: Optional[str] = None) -> None:
        self._cycle_id = uuid4().hex
        self._trace_id = trace_id or uuid4().hex
        self._status = CycleStatus.RUNNING
        self._phase = CyclePhase.INITIALIZING
        self._outcome: Optional[CycleOutcome] = None
        self._started_at = datetime.now(UTC)
        self._start_mono = time.monotonic()
        self._phase_start_mono = self._start_mono
        self._completed_at: Optional[datetime] = None
        self._error: Optional[str] = None
        self._error_stage: Optional[str] = None
        self._phase_durations: dict[str, float] = {}

    @property
    def cycle_id(self) -> str:
        return self._cycle_id

    @property
    def trace_id(self) -> str:
        return self._trace_id

    @property
    def phase(self) -> CyclePhase:
        return self._phase

    @property
    def status(self) -> CycleStatus:
        return self._status

    @property
    def elapsed_ms(self) -> float:
        return (time.monotonic() - self._start_mono) * 1000

    def transition_to(self, phase: CyclePhase) -> None:
        """Record phase transition and duration of the previous phase."""
        now = time.monotonic()
        prev_duration_ms = (now - self._phase_start_mono) * 1000
        self._phase_durations[self._phase.value] = round(prev_duration_ms, 1)
        self._phase = phase
        self._phase_start_mono = now

    def complete(
        self,
        outcome: CycleOutcome,
    ) -> None:
        """Mark cycle as completed with the given outcome."""
        now = time.monotonic()
        prev_duration_ms = (now - self._phase_start_mono) * 1000
        self._phase_durations[self._phase.value] = round(prev_duration_ms, 1)

        self._phase = CyclePhase.COMPLETED
        self._status = CycleStatus.COMPLETED
        self._outcome = outcome
        self._completed_at = datetime.now(UTC)

    def fail(
        self,
        error: str,
        *,
        stage: Optional[str] = None,
        timed_out: bool = False,
    ) -> None:
        """Mark cycle as failed."""
        now = time.monotonic()
        prev_duration_ms = (now - self._phase_start_mono) * 1000
        self._phase_durations[self._phase.value] = round(prev_duration_ms, 1)

        self._phase = CyclePhase.FAILED
        self._status = CycleStatus.TIMED_OUT if timed_out else CycleStatus.FAILED
        self._outcome = CycleOutcome.PIPELINE_ERROR
        self._error = error
        self._error_stage = stage
        self._completed_at = datetime.now(UTC)

    def to_state(self) -> CycleState:
        """Snapshot the current cycle state."""
        return CycleState(
            cycle_id=self._cycle_id,
            trace_id=self._trace_id,
            status=self._status,
            phase=self._phase,
            outcome=self._outcome,
            started_at=self._started_at,
            phase_started_at=datetime.now(UTC),
            completed_at=self._completed_at,
            error=self._error,
            error_stage=self._error_stage,
            phase_durations_ms=dict(self._phase_durations),
        )
